### PR TITLE
Improve `multiprocessing.context` module

### PR DESCRIPTION
**Pull request by sobolevn**
Two main changes:
1. I think it is better to use assignment in case of alises where possible. For example, `Process` is simply defined as `Process = Process` in [cpython's `DefaultContext`](https://github.com/python/cpython/blob/75a6441718dcbc65d993c9544e67e25bef120e82/Lib/multiprocessing/context.py#L231). This is more accurate: it is set as an explicit classvar and we have full type info (`type[Process]` is wider)
2. There's no need to redefine `SpawnContext` and `SpawnProcess` in `if`. They are the same in case of typing. Only implementation details differ: 
  - https://github.com/python/cpython/blob/75a6441718dcbc65d993c9544e67e25bef120e82/Lib/multiprocessing/context.py#L283-L288
  - https://github.com/python/cpython/blob/75a6441718dcbc65d993c9544e67e25bef120e82/Lib/multiprocessing/context.py#L331-L336

### CODE REVIEW

**Comment by sobolevn** [comment]
Oh, looks like we have a problem with `pyright`:
>  /home/runner/work/typeshed/typeshed/stdlib/multiprocessing/context.pyi:29:20 - error: Type alias "ProcessError" cannot use itself in its definition (reportGeneralTypeIssues)

**Comment by AlexWaygood** [comment]
> 1. This is more accurate: it is set as an explicit classvar

FYI, there's actually an annoying mypy bug that means that assignments inside class bodies are interpreted as instance attributes rather than `ClassVar`s. This bug definitely exists for `.pyi` files, and might also exist for `.py` files. Not sure if there's an open bug report about it or not. 

And yeah, there's the pyright bug w.r.t. variable scope, that you've already identified. We've come across it a few times now (see `multiprocessing.reduction` for some hacky workarounds). Someone should probably file a bug report for that, too 😅

File: stdlib/multiprocessing/context.pyi [hunk 1]
```diff
@@ -27,11 +26,11 @@ class TimeoutError(ProcessError): ...
 class AuthenticationError(ProcessError): ...
 
 class BaseContext:
-    Process: type[BaseProcess]
-    ProcessError: type[Exception]
-    BufferTooShort: type[Exception]
-    TimeoutError: type[Exception]
-    AuthenticationError: type[Exception]
+    Process: ClassVar[type[BaseProcess]]
```
Comment (srittau) [inline stdlib/multiprocessing/context.pyi:29]: I don't see this field, at least in Python 3.9.

File: stdlib/multiprocessing/context.pyi [hunk 2]
```diff
@@ -147,43 +146,33 @@ class DefaultContext(BaseContext):
 
 _default_context: DefaultContext
 
+class SpawnProcess(BaseProcess):
+    _start_method: str
+    @staticmethod
+    def _Popen(process_obj: BaseProcess) -> SpawnProcess: ...
```
Comment (srittau) [inline stdlib/multiprocessing/context.pyi:152]: I know you just moved this, but this returns `multiprocessing.popen.Popen`.

**Review by srittau — CHANGES_REQUESTED** [review]

File: stdlib/multiprocessing/context.pyi (hunk 1 above)
Comment (sobolevn) [reply stdlib/multiprocessing/context.pyi:29]: Hm, it exists here: https://github.com/python/typeshed/blob/df4618ef2dec1d27e363da0fc4ba8b00a2c811fe/stdlib/multiprocessing/context.pyi#L30

I just added a `ClassVar` here 🤔

Comment (srittau) [reply stdlib/multiprocessing/context.pyi:29]: Doesn't exist in HEAD either: https://github.com/python/cpython/blob/75a6441718dcbc65d993c9544e67e25bef120e82/Lib/multiprocessing/context.py#L30

Let's just remove it.

File: stdlib/multiprocessing/context.pyi [hunk 3]
```diff
@@ -27,11 +26,11 @@ class TimeoutError(ProcessError): ...
...
-    Process: type[BaseProcess]
-    ProcessError: type[Exception]
-    BufferTooShort: type[Exception]
-    TimeoutError: type[Exception]
-    AuthenticationError: type[Exception]
+    Process: ClassVar[type[BaseProcess]]
+    ProcessError: ClassVar[type[Exception]]
+    BufferTooShort: ClassVar[type[Exception]]
+    TimeoutError: ClassVar[type[Exception]]
+    AuthenticationError: ClassVar[type[Exception]]
```
Comment (AlexWaygood) [inline stdlib/multiprocessing/context.pyi:33]: pyright is fine with this -- the pyright variable-scope bug only exists for assignments, not attribute annotations

```suggestion
    ProcessError: ClassVar[type[ProcessError]]
    BufferTooShort: ClassVar[type[BufferTooShort]]
    TimeoutError: ClassVar[type[TimeoutError]]
    AuthenticationError: ClassVar[type[AuthenticationError]]
```

File: stdlib/multiprocessing/context.pyi [hunk 4]
```diff
@@ -146,33 +145,43 @@ class DefaultContext(BaseContext):
...
         Process: ClassVar[type[ForkServerProcess]]
+else:
+    class SpawnProcess(BaseProcess):
+        _start_method: str
+        @staticmethod
+        def _Popen(process_obj: BaseProcess) -> popen_spawn_win32.Popen: ...
+
+    class SpawnContext(BaseContext):
+        _name: str
+        Process: ClassVar[type[SpawnProcess]]
```
Comment (AlexWaygood) [inline stdlib/multiprocessing/context.pyi:183]: Can we merge these class definitions with the ones in the `!= win32"` branch? Only the return type of `SpawnProcess._Popen` seems to be different; I think we can just do a conditional definition of that one staticmethod inside the class.

File: stdlib/multiprocessing/context.pyi [hunk 5]
```diff
@@ -1,13 +1,11 @@
 import ctypes
-import multiprocessing
 import sys
 from collections.abc import Callable, Iterable, Sequence
 from ctypes import _CData
 from logging import Logger
-from multiprocessing import queues, synchronize
+from multiprocessing import popen_fork, popen_forkserver, popen_spawn_posix, popen_spawn_win32, queues, synchronize
 from multiprocessing.connection import _ConnectionBase
 from multiprocessing.managers import SyncManager
```
Comment (AlexWaygood) [inline stdlib/multiprocessing/context.pyi:8]:
```suggestion
from multiprocessing.managers import SyncManager
from multiprocessing.pool import Pool as _Pool
```

File: stdlib/multiprocessing/context.pyi (hunk 4 above)
Comment (AlexWaygood) [reply stdlib/multiprocessing/context.pyi]: But this can wait for another PR.

**Review by AlexWaygood — APPROVED** [review]

File: stdlib/multiprocessing/context.pyi (hunk 4 above)
Comment (sobolevn) [reply stdlib/multiprocessing/context.pyi]: Yes, I can! I'll send a new one tomorrow, any reasonable working hours are ending in my part of Russia 🙂

**Review by srittau — APPROVED** [review]

**Pull request merged**